### PR TITLE
string-ot-v1

### PR DESCRIPTION
**Pull request by geyang**

### CODE REVIEW

File: workspace/package.json [hunk 1]
```diff
@@ -22,6 +22,7 @@
   "author": "Jiaming Guo, Cha Chen, Ge Yang",
   "license": "UNLICENSED",
   "devDependencies": {
+    "babel-cli": "^6.23.0",
```
Comment (geyang) [inline workspace/package.json]: babel cli should not be inside devDependency. @jiamingguo

File: workspace/src/string-transform.js [hunk 2]
```diff
@@ -38,13 +38,139 @@ function transform_ins_ins(op0, op1) {
...
+
+  // case 1: destination of Mov located in the deletion area
+  // case 1.1:
+  //     ______________
+  //    |              |
+  //    |              v
+  //   [_____]     [--------]
+  //  0 1 2 3 4 5 6 7 8 9 10 11 12 13 ...
+  //    op1           op0(del)
+
```
Comment (geyang) [inline workspace/src/string-transform.js]: These graphs are awesome! don't get carried away though~

File: workspace/src/string-transform.js [hunk 3]
```diff
@@ -38,13 +38,139 @@ function transform_ins_ins(op0, op1) {
...
 function transform_del_del(op0, op1) {
   return op1.pos <= op0.pos ? [op1] :
     op1.pos <= op0.pos + op0.length ? [makeOpDel(op0.pos, op1.value)] :
-    [makeOpDel(op1.pos - op0.length, op1.vlue)];
+    [makeOpDel(op1.pos - op0.length, op1.value)];
+}
+
+function transform_del_mov(op0, op1) {
+  // case 0: destination of Mov located before deletion area
+  if((op1.pos+op1.length<=op0.pos && op1.des <= op0.pos)||(op1.pos >= op0.pos+op0.length && (op1.des >= op0.pos+op0.length || op1.des <= op0.pos))){
```
Comment (geyang) [inline workspace/src/string-transform.js]: These lines are way too long. Limit lines to less than 80 characters

File: workspace/src/string-transform.js [hunk 4]
```diff
@@ -38,13 +38,139 @@ function transform_ins_ins(op0, op1) {
...
+  }
+
+  // case 3.3:
+  //     ___________________
+  //    |                   |
+  //    |                   V
+  //   {____[===}------]
+  //  0 1 2 3 4 5 6 7 8 9 10 11 12 13 ...
+  //    op1       op0(del)
+  if (op1.pos<=op0.pos && op1.pos+op1.length<=op0.pos+op0.length&& op1.pos+op1.length >= op0.pos && op1.des>=op0.pos+op0.length){
```
Comment (geyang) [inline workspace/src/string-transform.js]: all of these logics can be combined. 

1. Use single layer of `if` `else if`
2. combine cases
3. maybe assign some of the conditions to a variable.

File: workspace/src/utility.js [hunk 5]
```diff
@@ -14,11 +14,11 @@ export function makeOpDel (pos, length) {
   };
 };
 
-export function makeOpMov (pos, value, des) {
+export function makeOpMov (pos, length, des) {
   return {
     type : "mov",
     pos : pos,
-    value : value,
+    value : length,
```
Comment (geyang) [inline workspace/src/utility.js]: this shouldn't be `value: length` right? @jam-world @jiamingguo

**Review by geyang — COMMENTED** [review]
- logic need to be combined 
- `babel-cli` need to be removed from package.json.

File: workspace/src/utility.js (hunk 5 above)
Comment (symboli-artoria17) [reply workspace/src/utility.js]: I changed it to length. Because I found length is easier to work with.

Comment (geyang) [reply workspace/src/utility.js]: The problem is not length. The problem is that you are still using `value` as the key.

Comment (geyang) [reply workspace/src/utility.js]: I think it should be value: value.

@jam-world in `es6`, you do `{value}` and it is the same as `value: value`.

Comment (symboli-artoria17) [reply workspace/src/utility.js]: oh boy. Your are right. I'm correcting it to length:length

Comment (geyang) [reply workspace/src/utility.js]: @jiamingguo did you see my comment on `ES6` syntax?

File: workspace/src/string-ops.js [hunk 6]
```diff
@@ -7,6 +7,16 @@ export function del(s, i, l) {
 }
 
 export function mov(s, i, l, d) {
-    return s.slice(0, i) + s.slice(i+l, d) + s.slice(d, l);
+  if (d>=i+l){
```
Comment (geyang) [inline workspace/src/string-ops.js]: Format this code with IntelliJ/webstorm to comply with standard. (Add spaces)

File: workspace/src/string-transform.js [hunk 7]
```diff
@@ -38,13 +38,207 @@ function transform_ins_ins(op0, op1) {
...
+     // case 2.3:
+     //  __
+     // |  |
+     // |  |
+     // v {____[===}------]
+     //  0 1 2 3 4 5 6 7 8 9 10 11 12 13 ...
+     //    op1        op0(del)
+
+     */
+    if (op1.des >= op1.pos + l1 && op1.des <= op0.pos + l0) {
```
Comment (geyang) [inline workspace/src/string-transform.js]: Take a look at the transformIndex function. Can these logic be packaged and reused that function?

File: workspace/src/string-transform.js [hunk 8]
```diff
@@ -60,18 +254,17 @@ export function transform(op0, op1) {
...
-    };
+    }
   } else if (op0.type === "del") {
     if (op1.type === "ins") {
       return transform_del_ins(op0, op1);
     } else if (op1.type === "del") {
       return transform_del_del(op0, op1);
     } else if (op1.type === "mov") {
-      // todo: del then mov
-      return nil;
```
Comment (geyang) [inline workspace/src/string-transform.js]: @jam-world nil is not a JavaScript token ha~

File: workspace/package.json (hunk 1 above)
Comment (geyang) [reply workspace/package.json]: @jiaming please address this comment!

File: workspace/src/string-transform.js [hunk 9]
```diff
@@ -53,25 +247,24 @@ export function transform(op0, op1) {
   if (op0.type === "ins") {
```
Comment (geyang) [inline workspace/src/string-transform.js]: you could potentially use a dictionary to realize this function. @jiamingguo @jam-world 

Use 
```
transforms = {'ins,del': transform_ins_del, ...}
```

**Review by geyang — COMMENTED** [review]
looks good to me!

**Comment by jiamingguo** [comment]
testcases

**Review by symboli-artoria17 — COMMENTED** [review]
testcases

File: workspace/src/string-transform.js [hunk 10]
```diff
@@ -20,31 +20,335 @@ export function transformCursor(cursor, op) {
...
 function transform_del_del(op0, op1) {
   return op1.pos <= op0.pos ? [op1] :
     op1.pos <= op0.pos + op0.length ? [makeOpDel(op0.pos, op1.value)] :
-    [makeOpDel(op1.pos - op0.length, op1.vlue)];
+      [makeOpDel(op1.pos - op0.length, op1.value)];
+}
+
+function transform_del_mov(op0, op1) {
+  var l1 = op1.length;
+  var l0 = op0.length;
```
Comment (geyang) [inline workspace/src/string-transform.js]: should use `const` for all of these `var`s

File: workspace/src/string-transform.js [hunk 11]
```diff
@@ -20,31 +20,335 @@ export function transformCursor(cursor, op) {
...
+  }
+}
+
+function transform_ins_mov(op0, op1) {
+  var v0 = op0.value;
+  var l0 = v0.length;
+  var l1 = op1.length;
+  var des = op1.des;
+  var pos0 = op0.pos;
+  var pos1 = op1.pos;
```
Comment (geyang) [inline workspace/src/string-transform.js]: should use `const` for all of these `var`s

File: workspace/src/string-transform.js (hunk 9 above)
Comment (geyang) [reply workspace/src/string-transform.js]: @jiamingguo can you see if you can try this?

File: workspace/src/utility.js [hunk 12]
```diff
@@ -1,24 +1,24 @@
-export function makeOpIns (pos, value) {
+export function makeOpIns(pos, value) {
```
Comment (geyang) [inline workspace/src/utility.js]: we can rename these to `Ins`, `Del`, `Mov`.

Comment (symboli-artoria17) [reply workspace/src/utility.js]: Is it necessary to change? That function name is good though.

File: workspace/src/string-transform.js (hunk 11 above)
Comment (symboli-artoria17) [reply workspace/src/string-transform.js]: Just did it. See next commit

File: workspace/src/utility.js (hunk 12 above)
Comment (geyang) [reply workspace/src/utility.js]: If it is only used in test we don't have to change it. It does feel a bit redundant.

**Comment by episodeyang** [comment]
![screenshot 2017-03-17 00 25 20](https://cloud.githubusercontent.com/assets/630490/24029988/44fa500c-0aa8-11e7-9158-3a15a2b4c7a7.png）

ops test have not passed, so the transform code is likely to be incorrect as well. @jiamingguo

**Comment by geyang** [comment]
screenshot 2017-03-17 00 25 20

ops test have not passed, so the transform code is likely to be incorrect as well. @jiamingguo

**Comment by geyang** [comment]
@jiamingguo You tests are okay! those failed tests came from the `dist` folder. I have updated the npm test script to avoid those folders.

File: workspace/src/string-ops.js [hunk 13]
```diff
@@ -2,7 +2,7 @@ export function ins(s, i, v) {
   return s.slice(0, i) + v + s.slice(i);
 }
 
-export function del(s, i, l) {
+export function del(s, i, l = 1) {
```
Comment (geyang) [inline workspace/src/string-ops.js]: @jam-world Otherwise the slice i + l would return `NaN`.

Comment (symboli-artoria17) [reply workspace/src/string-ops.js]: I will add a new test case for this.

Comment (geyang) [reply workspace/src/string-ops.js]: that is just the default for `l`.

**Comment by geyang** [comment]
Looks pretty good to me, we can probably merge this. 

Milestone `string-ot-v1` complete? @jiamingguo

**Comment by geyang** [comment]
this one is ready to be merged. @jiamingguo

**Pull request merged**